cmd/gardenlet/app/app.go

```diff
@@ -480,6 +489,142 @@ func (g *garden) cleanupOrphanedExtensionsServiceAccounts(ctx context.Context, g
+				return fmt.Errorf("failed to recreate the original secret %w", err)
+			}
+		}
+
+		// secret was already recreated. just delete the temporary one
+		if err := c.Delete(ctx, &temp); client.IgnoreNotFound(err) != nil {
+			return err
+		}
+	}
+
+	secretsToRecreate, err := getSecretsToRecreate(ctx, c)
+	if err != nil {
+		return fmt.Errorf("failed listing secrets for recreation %w", err)
+	}
+
+	for _, original := range secretsToRecreate {
+		tempSecret := original.DeepCopy()
+		tempSecret.Name = original.Name + "-tmp"
+		metav1.SetMetaDataLabel(&tempSecret.ObjectMeta, tempSecretLabel, "true")
+		metav1.SetMetaDataLabel(&tempSecret.ObjectMeta, tempSecretOldNameLabel, original.Name)
+		tempSecret.DeletionTimestamp = nil
+		tempSecret.ResourceVersion = ""
+		tempSecret.Finalizers = []string{}
+
+		if err := c.Create(ctx, tempSecret); err != nil {
+			return fmt.Errorf("failed to create a temporary secret %w", err)
+		}
+
+		if err := removeFinalizersAndWait(ctx, c, original.DeepCopy()); err != nil {
+			return err
+		}
+
+		// zero meta info
+		original.DeletionTimestamp = nil
+		original.ResourceVersion = ""
+		original.Finalizers = []string{}
+
+		// recreate the original and delete the temporary one
+		if err := c.Create(ctx, &original); err != nil {
+			return fmt.Errorf("failed to recreate the original secret %w", err)
+		}
+		if err := c.Delete(ctx, tempSecret); client.IgnoreNotFound(err) != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// TODO(dimityrmirchev): Remove this code after v1.84 has been released.
+func removeFinalizersAndWait(ctx context.Context, c client.Client, secret *corev1.Secret) error {
+	patch := client.StrategicMergeFrom(secret.DeepCopy())
+	secret.Finalizers = []string{}
+	if err := c.Patch(ctx, secret, patch); err != nil {
+		return fmt.Errorf("failed to patch the original secret %w", err)
+	}
+
+	goneSecret := &corev1.Secret{ObjectMeta: metav1.ObjectMeta{Name: secret.Name, Namespace: secret.Namespace}}
+	cancelCtx, cancel := context.WithTimeout(ctx, time.Second*10)
+	defer cancel()
+	return kubernetesutils.WaitUntilResourceDeleted(cancelCtx, c, goneSecret, 1*time.Second)
+}
+
+// TODO(dimityrmirchev): Remove this code after v1.84 has been released.
+func getSecretsToRecreate(ctx context.Context, c client.Client) ([]corev1.Secret, error) {
+	selector := labels.NewSelector()
+	isGC, err := labels.NewRequirement(references.LabelKeyGarbageCollectable, selection.Equals, []string{"true"})
+	if err != nil {
+		return nil, err
+	}
+	notTemp, err := labels.NewRequirement(tempSecretLabel, selection.DoesNotExist, []string{})
+	if err != nil {
+		return nil, err
+	}
+	selector.Add(*isGC, *notTemp)
+	secretList := &corev1.SecretList{}
+	if err := c.List(ctx, secretList, client.MatchingLabelsSelector{Selector: selector}); err != nil {
+		return nil, err
+	}
+	secretsToRecreate := slices.DeleteFunc(secretList.Items, func(s corev1.Secret) bool {
+		return !slices.Contains(s.Finalizers, grmFinalizer)
```

Comment:
What about the deletion timestamp? Aren't you now also considering secrets which DON'T have a deletion timestamp but just the finalizer (which is perfectly fine)?

Reply:
Yes, you are right. Should be fixed with https://github.com/gardener/gardener/pull/8812/commits/0d25953b938b04a0c37e401f6e286efb5fd16a67